### PR TITLE
[templates] Align Android back button behavior with classic build

### LINKED ISSUE #15326: EAS build vs expo build: code executing differences, REPOs with 2 demos of bug (images and fonts)
### Summary

I tried both build systems and behavior of “back” button in android is different:

In old classic expo build when I press “back” app is not really closed, it is running in background, when I run app again it is opened immediately.

In new EAS build when I press “back” button and run it again (all with same 2 seconds delay) I see splash loading screen which stays visible much less time than when app is really closed (“cold” start) but splash screen means that app is loading again (!) but app was not really (fully) closed. Some code of app works like it is cold start (inits) and some code (mostly animations with “useNativeDriver”) is still running.

This looks very strange.

Are big differences in handling “back” button in these build systems and why on new EAS build I see splash startup screen every time I open app even after 2 seconds? On old expo build system the same code base works as expected (no splash screen on reopen app)

Original discussion is here https://forums.expo.dev/t/eas-build-vs-expo-build-code-executing-differences/59384

The same problem has other users. Such strange behavior is in any project.

--------------

As expo team asked for working examples, I did it. How it was:
---

I just started clear new empty project on SDK 43 and EAS 0.37. With default template demo code all works same as in classic expo build, so I thought maybe problem is elsewhere...

I look how my apps are starting and what can show splash screen and found that I use `expo-asset` module which awaits for assets to be loaded and during that it shows default loader via `expo-app-loading` module.

I caught this bug and made working example. The same code, but different behavior when I press "back" button on android and then after 2 seconds (no matter really how much time) I open app again:

- With classic expo build: app is showing immediately with images, no splash screen visible

- With EAS build: app is quickly showing splash screen (that means assets are loaded again! but WHY?!) and then app is showing and then images are appearing like it is first time "cold" start - every image is loaded with little delay)

So... that means EAS build AND/OR `expo-asset` module produce bug(s). And it is really visible.

P.S. Later I tested this with fonts, same behavior as with assets. See below 2 bug demo examples.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

42 and 43

### Environment

SDK: 42 and 43
eas: 0.37.0
node js: v16.6.1
npm: 7.20.3
os: win 10

### Reproducible demo

I caught and tested this with one image, but for better visible effect I added example with 3 big images.

Bug demo # 1 with images
---

Repo with reproducible bug and how to test instructions:
https://github.com/psnet/eastest

Recorded video of this bug with and slowed it 2 times for better visual checking:
https://www.youtube.com/watch?v=9PaCMu4TOPU

Bug demo # 2 with fonts
---

Repo with reproducible bug and how to test instructions:
https://github.com/psnet/eastest2

Recorded video of this bug with and slowed it 2 times for better visual checking:
https://www.youtube.com/watch?v=rkgFkDZVcEo

### PR DESCRIPTION
**Pull request by Kudo**
# Why

fix #15326 where we have the default Android implementation on eas build but move activities to background on [classic build](https://github.com/expo/expo/blob/28d5976df2379540936083e599b100a50bd3575d/android/expoview/src/main/java/host/exp/exponent/experience/BaseExperienceActivity.kt#L93-L103). actually, finishing activities is the default behavior on Android, so i kinda dunno which solution is better. after i found that on Android S, [the default behavior for root activities is also moving background](https://developer.android.com/about/versions/12/behavior-changes-all#back-press), i would try to align the design with Android S.

# How

- for android <= 11 + root activities, move to background.
- for android <= 11 + non-root activities, use the platform default implementation (`finish()`)
- for android > 11, use the platform default implementation. (assuming to move root activities in background) 

# Test Plan

use the [example repo](https://github.com/psnet/eastest) from #15326 + this patch
- test `expo run:android`
- test `eas build`

# Checklist

- n/a Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- n/a This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: templates/expo-template-bare-minimum/android/app/src/main/java/com/helloworld/MainActivity.java [hunk 1]
```diff
@@ -39,4 +39,24 @@ protected ReactRootView createRootView() {
...
   }
+
+  /**
+   * We try to align the back button behavior with Android S
+   * where moving root activities to background instead of finishing activities.
+   * @see <a href="https://developer.android.com/reference/android/app/Activity#onBackPressed()">onBackPressed</a>
+   */
+  @Override
+  public void invokeDefaultOnBackPressed() {
+    if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.R) {
```
Comment (brentvatne) [inline templates/expo-template-bare-minimum/android/app/src/main/java/com/helloworld/MainActivity.java:50]: missing `import android.os.Build;`?

**Review by brentvatne — APPROVED** [review]

File: templates/expo-template-bare-minimum/android/app/src/main/java/com/helloworld/MainActivity.java [hunk 2]
```diff
@@ -39,4 +39,24 @@ protected ReactRootView createRootView() {
       }
     });
   }
+
+  /**
+   * We try to align the back button behavior with Android S
```
Comment (EvanBacon) [inline templates/expo-template-bare-minimum/android/app/src/main/java/com/helloworld/MainActivity.java:44]:
```suggestion
   * Align the back button behavior with Android S
```

**Comment by psnet** [comment]
`templates/expo-template-bare-minimum` is using during `EAS build`?

**Comment by Kudo** [comment]
@psnet exactly! and we try to maintain the template as minimal as possible. that's one reason of `eas build` generated apps' size getting smaller than classic `expo build`.

**Pull request merged**